### PR TITLE
Add no_std embedded test in CI and fix for no_std

### PR DESCRIPTION
**Pull request by RCasatta**
Introduce a test for no_std launched by the CI through qemu.

Firstly the test is failing due to the way `core` is exported, d1cfc2c33dc4de479b7ca21baf50e055436d2bb4 fix that

Note: to use this version in rust-bitcoin, `extern crate core` is needed in rust-bitcoin

### CODE REVIEW

**Comment by apoelstra** [comment]
Neat! Can you reorder the commits so that they both pass during bisection?

**Comment by RCasatta** [comment]
> Neat! Can you reorder the commits so that they both pass during bisection?

yes, done

**Comment by RCasatta** [comment]
I force pushed a second time because of the apt error in CI that happened a second time, will re-try in while

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -80,3 +80,25 @@ jobs:
           DO_FEATURE_MATRIX: true
         run: ./contrib/test.sh
 
+  Embedded:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Set up QEMU
+        run: sudo apt install qemu-system-arm
```
Comment (TheBlueMatt) [inline .github/workflows/rust.yml:89]: You usually want to run an `apt update` or `apt-get update` immediately before an install to make sure you're pulling the latest versions (which is why you're getting 404s in CI now).

Comment (RCasatta) [reply .github/workflows/rust.yml]: right, thanks

**Review by TheBlueMatt — APPROVED** [review]
I admit I don't know what half the embedded things here do, but they seem to be actually running correctly in CI and the main patch looks good.

**Review by apoelstra — APPROVED** [review]
utACK.

Agree with Matt that I don't really understand how this test works but it looks right. Good call fixing CI 404s - I just kicked it :)

File: src/lib.rs [hunk 2]
```diff
@@ -37,7 +37,7 @@
 #![cfg_attr(all(test, feature = "unstable"), feature(test))]
 #[cfg(all(test, feature = "unstable"))] extern crate test;
 
-#[cfg(any(test, feature="std"))] pub extern crate core;
+#[cfg(any(test, feature="std"))] extern crate core;
```
Comment (sgeisler) [inline src/lib.rs:40]: This is technically a breaking change, although it seems to have been a bug in the first place to re-export core.

**Comment by sgeisler** [comment]
I'm a bit unhappy that this was merged so quickly now that CI finally worked (I typically ignore PRs as long as they have failing CI). We should keep in mind that removing the core re-export is breaking. It's dumb for anyone to use it, but IntelliJ Rust for example has used it to auto-import stuff from std for me in the past. If someone did not notice and kept it that way that might be a problem.

**Comment by RCasatta** [comment]
Probably by luck, and in the end it's better that the lack of apt update has been exposed, but the CI was working before the commit reordering https://github.com/rust-bitcoin/bitcoin_hashes/actions/runs/751200720

**Comment by sgeisler** [comment]
Oh, I see, then sorry for not reviewing earlier. Just saw it today in the broken state and restarted the tests to come back later :grimacing:

**Comment by apoelstra** [comment]
@sgeisler I merged it so quickly only because it looked pretty-much trivial.

It did occur to me that the core re-export is technically breaking but I find it really hard to imagine it actually breaking anything in practice and I don't think we should bother doing a major version bump for it.

**Comment by sgeisler** [comment]
I agree that normally it might be worth an exception to SemVer, but with auto importers being stupid at times I can totally see it break someone's project. We will see I guess :sweat_smile:

**Comment by TheBlueMatt** [comment]
This broke building rust-bitcoin, which now fails with eg
```
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> /home/matt/.cargo/registry/src/github.com-1ecc6299db9ec823/bitcoin-0.26.0/src/hash_types.rs:54:1
   |
54 | hash_newtype!(FilterHeader, sha256d::Hash, 32, doc="Filter header, as defined in BIP-157");
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ maybe a missing crate `core`?
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Comment by RCasatta** [comment]
I wrote in the PR description this would need extern crate core in rust-bitcoin

**Pull request merged**